lib/internal/process/promises.js

```diff
@@ -41,88 +41,179 @@ const { isErrorStackTraceLimitWritable } = require('internal/errors');
 // *Must* match Environment::TickInfo::Fields in src/env.h.
 const kHasRejectionToWarn = 1;
 
-const maybeUnhandledPromises = new SafeWeakMap();
-const pendingUnhandledRejections = [];
-const asyncHandledRejections = [];
-let lastPromiseId = 0;
+// By default true because in cases where process is not a global
+// it is not possible to determine if the user has added a listener
+// to the process object.
+let hasMultipleResolvesListener = true;
 
-// --unhandled-rejections=none:
-// Emit 'unhandledRejection', but do not emit any warning.
-const kIgnoreUnhandledRejections = 0;
+if (process.on) {
+  hasMultipleResolvesListener = process.listenerCount('multipleResolves') !== 0;
 
-// --unhandled-rejections=warn:
-// Emit 'unhandledRejection', then emit 'UnhandledPromiseRejectionWarning'.
-const kAlwaysWarnUnhandledRejections = 1;
+  process.on('newListener', (eventName) => {
+    if (eventName === 'multipleResolves') {
+      hasMultipleResolvesListener = true;
+      resolveErrorResolve = resolveErrorResolveHandler;
+      resolveErrorReject = resolveErrorRejectHandler;
+    }
+  });
 
-// --unhandled-rejections=strict:
-// Emit 'uncaughtException'. If it's not handled, print the error to stderr
-// and exit the process.
-// Otherwise, emit 'unhandledRejection'. If 'unhandledRejection' is not
-// handled, emit 'UnhandledPromiseRejectionWarning'.
-const kStrictUnhandledRejections = 2;
+  process.on('removeListener', (eventName) => {
+    if (eventName === 'multipleResolves') {
+      hasMultipleResolvesListener = process.listenerCount('multipleResolves') !== 0;
+      if (!hasMultipleResolvesListener) {
+        resolveErrorResolve = noop;
+        resolveErrorReject = noop;
+      }
+    }
+  });
+}
 
-// --unhandled-rejections=throw:
-// Emit 'unhandledRejection', if it's unhandled, emit
-// 'uncaughtException'. If it's not handled, print the error to stderr
-// and exit the process.
-const kThrowUnhandledRejections = 3;
+/**
+ * Errors & Warnings
+ */
+
+class UnhandledPromiseRejection extends Error {
+  code = 'ERR_UNHANDLED_REJECTION';
+  name = 'UnhandledPromiseRejection';
+  /**
+   * @param {Error} reason
+   */
+  constructor(reason) {
+    const message = 'This error originated either by throwing inside of an ' +
+    'async function without a catch block, or by rejecting a promise which ' +
+    'was not handled with .catch(). The promise rejected with the reason "' +
+    noSideEffectsToString(reason) + '".';
+
+    if (isErrorStackTraceLimitWritable()) {
+      const stackTraceLimit = Error.stackTraceLimit;
+      super(message);
+      Error.stackTraceLimit = stackTraceLimit;
```

Comment:
Saving the `stackTraceLimit` and setting it again after the super call should not have any impact in case `Error.prototype.constructor()` was not monkey patched. I guess the intention was to set the limit to zero before the call to `super()`?
The same below.

Reply:
uff. yes

Reply:
Please also address the same in the other error constructors

Reply:
done